pkg/provider/azure_routes.go

```diff
@@ -72,26 +72,26 @@ type delayedRouteUpdater struct {
 	interval time.Duration
 
 	lock           sync.Mutex
-	routesToUpdate []*delayedRouteOperation
+	routesToUpdate []batchOperation
 }
 
 // newDelayedRouteUpdater creates a new delayedRouteUpdater.
-func newDelayedRouteUpdater(az *Cloud, interval time.Duration) *delayedRouteUpdater {
+func newDelayedRouteUpdater(az *Cloud, interval time.Duration) batchProcessor {
 	return &delayedRouteUpdater{
 		az:             az,
 		interval:       interval,
-		routesToUpdate: make([]*delayedRouteOperation, 0),
+		routesToUpdate: make([]batchOperation, 0),
 	}
 }
 
 // run starts the updater reconciling loop.
-func (d *delayedRouteUpdater) run() {
-	err := wait.PollImmediateInfinite(d.interval, func() (bool, error) {
+func (d *delayedRouteUpdater) run(ctx context.Context) {
+	err := wait.PollUntilContextCancel(ctx, d.interval, true, func(ctx context.Context) (bool, error) {
 		d.updateRoutes()
 		return false, nil
 	})
-	if err != nil { // this should never happen, if it does, panic
-		panic(err)
+	if err != nil {
+		klog.Infof("delayedRouteUpdater: stopped due to %s", err.Error())
```

Comment:
if this exits, how could we ensure it is restarted again?

Reply:
We create a route updater gorouting each time the cloud provider restarts. This PR ensures the old one will be deleted because when it restarts, the ctx will be cancelled. A new one will be created when the new service controller initializes. I am testing this behavior and not ready for review yet.

Reply:
Ready for review now.